drivers/gpu/drm/vc4/vc4_hdmi.c

```diff
@@ -689,6 +689,30 @@ static void vc4_hdmi_set_spd_infoframe(struct drm_encoder *encoder)
 	vc4_hdmi_write_infoframe(encoder, &frame);
 }
 
+static void vc4_hdmi_set_hdmi_infoframe(struct drm_encoder *encoder)
+{
+	struct vc4_hdmi *vc4_hdmi = encoder_to_vc4_hdmi(encoder);
+	struct drm_connector *connector = &vc4_hdmi->connector;
+	struct drm_crtc *crtc = encoder->crtc;
+	const struct drm_display_mode *mode = &crtc->state->adjusted_mode;
+	union hdmi_infoframe frame;
+	int ret;
+
+	ret = drm_hdmi_vendor_infoframe_from_display_mode(&frame.vendor.hdmi,
+							  connector,
+							  mode);
+	if (ret < 0) {
+		DRM_ERROR("couldn't fill HDMI infoframe (%d)\n", ret);
+		return;
+	}
+	ret = hdmi_vendor_infoframe_check(&frame.vendor.hdmi);
```

Comment:
Seeing as we've only just generated this, is there a possibility that it is wrong? It seems to vary between drivers whether they check this or not.

Reply:
I was looking at intel_hdmi which does do this (for this infoframe and others).
Happy to skip it if it's believed redundant at this point.

I also noticed we aren't adding drm_hdmi_infoframe_set_hdr_metadata which we may want in future
(although I suspect Maxime may be adding this in 10-bit colour work.).

Reply:
i915 is a pretty good reference :-)
Even so, I think it's impossible.
Looking at ```hdmi_vendor_infoframe_check_only``` - https://elixir.bootlin.com/linux/latest/source/drivers/video/hdmi.c#L515, 
```type```, ```version```, and ```oui``` will have been set by ```hdmi_vendor_infoframe_init```
```frame->length``` has been set to ```hdmi_vendor_infoframe_length(frame)```
That only leaves ```frame->vic != 0 && frame->s3d_struct != HDMI_3D_STRUCTURE_INVALID```. 
```frame->vic``` is set by ```drm_mode_hdmi_vic``` which includes ```if (mode->flags & DRM_MODE_FLAG_3D_MASK) return 0;```, so that should ensure it is 0 in any 3D mode.

HDR patches are in progress, but yes that is the right patchset to be adding the infoframe. There's no point in sending it if you don't support HDR.